Fix issue #12034: add autofixes for unnecessary_fallible_conversions

roife opened the pull request:
fixes #12034

Currently, the `unnecessary_fallible_conversions` lint was capable of autofixing expressions like `0i32.try_into().unwrap()`. However, it couldn't autofix expressions in the form of `i64::try_from(0i32).unwrap()` or `<i64 as TryFrom<i32>>::try_from(0).unwrap()`.

This pull request extends the functionality to correctly autofix these latter forms as well.

changelog: [`unnecessary_fallible_conversions`]: Add autofixes for more forms

PartiallyTyped commented on tests/ui/unnecessary_fallible_conversions.rs:
```diff
@@ -3,4 +3,16 @@
     let _: i64 = 0i32.try_into().unwrap();
     let _: i64 = 0i32.try_into().expect("can't happen");
+
+    let _ = i64::try_from(0i32).unwrap();
```
Hi, thanks for contributing!

This is a slight nit, but could you add annotations to indicate whether that line should be linted? This helps avoid regressions when running `uibless` to account for new tests.

```suggestion
    let _ = i64::try_from(0i32).unwrap();
    //~^ ERROR use of a fallible conversion when an infallible one could be used
```

There are some examples here: 
https://rustc-dev-guide.rust-lang.org/tests/ui.html#error-annotation-examples

And some details here:
https://rustc-dev-guide.rust-lang.org/tests/ui.html#error-annotations

roife replied:
Thank you for reminder! I have added the necessary error annotations and included some additional comments in the tests to make it clearer.

PartiallyTyped commented on clippy_lints/src/methods/unnecessary_fallible_conversions.rs:
```diff
@@ -45,47 +46,80 @@ fn check<'tcx>(
+            primary_span
+        };
+
+        let qpath_spans = qpath.and_then(|qpath| match qpath {
+            QPath::Resolved(_, path) => {
+                let segments = path.segments.iter().map(|seg| seg.ident).collect::<Vec<_>>();
+                (segments.len() == 2).then(|| vec![segments[0].span, segments[1].span])
+            },
+            QPath::TypeRelative(_, seg) => Some(vec![seg.ident.span]),
+            QPath::LangItem(_, _) => unreachable!("`TryFrom` and `TryInto` are not lang items"),
+        });
```
This is a bit of a nit, and may be unnecessary, but if you return `Option<[&'_ span]>`, I think the matching could be cleaner as you will be matching something like

```
(FunctionKind::TryFromFunction, [from_method_span], Some(unwrap_span)) => todo!(),
(FunctionKind::TryFromFunction, [from_cls_span, from_method_span], Some(unwrap_span)) => todo!(),
_ => unreachable!()
```

which I think would make this a bit cleaner, other than that, I think it's very good and a great addition!

roife replied:
Sorry it took me so long to see this comment. I tried this but it introduced too many arms (multiple `FunctionKind::TryFromFunction`s and `FunctionKind::TryIntoFunction`s) and made the code more complicated

PartiallyTyped replied:
I see. I am good then, but I think we still need another reviewer with r perms. Maybe zulip?

PartiallyTyped reviewed (approved):
Looks good to me, thank you for the auto-fixes!

I think we need to find another reviewer for this though 😅 

@rustbot r? clippy

Centri3 commented on clippy_lints/src/methods/unnecessary_fallible_conversions.rs:
```diff
@@ -29,6 +29,7 @@ fn check<'tcx>(
     node_args: ty::GenericArgsRef<'tcx>,
     kind: FunctionKind,
     primary_span: Span,
+    qpath: Option<&QPath<'_>>,
```
imo this should be stored in `FunctionKind`, in `TryFromFunction` and `TryIntoFunction`, as the path is only passed in those cases (and shouldn't be for a method call)

Centri3 commented on clippy_lints/src/methods/unnecessary_fallible_conversions.rs:
```diff
@@ -45,47 +46,80 @@ fn check<'tcx>(
+                };
+                (self_ty, other_ty, sugg, Applicability::MachineApplicable)
+            },
+            (FunctionKind::TryFromFunction, _, _) => {
+                let sugg = vec![(primary_span, String::from("From::from"))];
+                (other_ty, self_ty, sugg, Applicability::Unspecified)
+            },
+            (FunctionKind::TryIntoFunction, _, _) => {
+                let sugg = vec![(primary_span, String::from("Into::into"))];
+                (self_ty, other_ty, sugg, Applicability::Unspecified)
+            },
+            (FunctionKind::TryIntoMethod, _, _) => {
+                let sugg = vec![(primary_span, String::from("into"))];
+                (self_ty, other_ty, sugg, Applicability::Unspecified)
             },
```
Could we simplify this by adding a method to `FunctionKind` for the suggestion? Something like:
`(other_ty, self_ty, kind.sugg(), Applicability::Unspecified)`.

This applies to `self_ty` and `other_ty` as well, not bc of code duplication but just that it's hard to understand what they are (without scrolling up and seeing `source_ty` and `target_ty`). And the entire match should be extracted tbh

Overall this entire match is a bit of a mess, once I have my desktop back (probably this weekend) I'll take a closer look and try my hand at it. It's pretty hard to understand in its current form.

Centri3 commented on clippy_lints/src/methods/unnecessary_fallible_conversions.rs:
```diff
@@ -45,47 +46,80 @@ fn check<'tcx>(
+                (self_ty, other_ty, sugg, Applicability::MachineApplicable)
+            },
+            (FunctionKind::TryFromFunction, Some(spans), Some(unwrap_span)) => {
+                let sugg = match spans.len() {
+                    1 => vec![(spans[0], String::from("from")), (unwrap_span, String::new())],
+                    2 => vec![
+                        (spans[0], String::from("From")),
+                        (spans[1], String::from("from")),
+                        (unwrap_span, String::new()),
+                    ],
+                    _ => unreachable!(),
```
Can we use destructuring here? As an example:
```suggestion
                let sugg = match &*spans {
                    [first] => vec![(first, String::from("from")), (unwrap_span, String::new())],
                    [first, second] => vec![
                        (first, String::from("From")),
                        (second, String::from("from")),
                        (unwrap_span, String::new()),
                    ],
                    _ => unreachable!(),
```
If you want you could also use an enum to remove the wildcard branch, like: `OneSpan(_)` + `TwoSpans(_, _)`.

Centri3 commented on clippy_lints/src/methods/unnecessary_fallible_conversions.rs:
```diff
@@ -45,47 +46,80 @@ fn check<'tcx>(
+
+        let qpath_spans = qpath.and_then(|qpath| match qpath {
+            QPath::Resolved(_, path) => {
+                let segments = path.segments.iter().map(|seg| seg.ident).collect::<Vec<_>>();
+                (segments.len() == 2).then(|| vec![segments[0].span, segments[1].span])
```
Iterating then collecting into a `Vec` is not worth doing to not do a field access twice. This also turns our borrowed slice into an owned vector, which I don't think is needed? (Maybe I'm missing something about ownership here - can't check rn.)

Destructuring can be used here as well:
```suggestion
                if let [first, second] = path.segments {
                    Some(first.ident.span, second.ident.span)
                } else {
                    None
                }
```

Centri3 reviewed (commented):
I'll take a closer look once I can. For now I have a few things to suggest

roife commented:
> For now I have a few things to suggest

Thanks for your suggestions! I have refactored the code according to your advice: I removed the large match statement, split it into several parts, and separated them into various sub-functions, which made the logic of the code much clearer.

Centri3 commented on clippy_lints/src/methods/unnecessary_fallible_conversions.rs:
```diff
@@ -2,25 +2,79 @@ use clippy_utils::diagnostics::span_lint_and_then;
+    }
+
+    fn machine_applicable_sugg(&self, primary_span: Span, unwrap_span: Span) -> Vec<(Span, String)> {
+        use FunctionKind::*;
+        use SpansKind::*;
```
On top of the wildcard, don't do this (match enum variants or unit structs with a single path segment). If you remove the `TryIntoMethod` (or any unit variant), that will instead be a wildcard pattern binding to the ident "TryIntoMethod", instead of a compiler error

There's an open issue for a lint about this #11440

roife replied:
Thanks for the suggestion. How about using `use FunctionKind as F`?

Centri3 commented on clippy_lints/src/methods/unnecessary_fallible_conversions.rs:
```diff
@@ -2,25 +2,79 @@ use clippy_utils::diagnostics::span_lint_and_then;
+                TryFromFunction(_) => ("From", "from"),
+                TryIntoFunction(_) | TryIntoMethod => ("Into", "into"),
+            };
+            (a.to_string(), b.to_string())
```
nit: Should be `to_owned` (https://users.rust-lang.org/t/to-string-vs-to-owned-for-string-literals/1441/2)

Centri3 reviewed (approved):
Just 2 nits, once dogfood passes this is ready :3

Centri3 commented on clippy_lints/src/methods/unnecessary_fallible_conversions.rs:
```diff
@@ -2,25 +2,74 @@ use clippy_utils::diagnostics::span_lint_and_then;
+}
+
+impl FunctionKind {
+    fn applicability(&self, parent_unwrap_call: &Option<Span>) -> Applicability {
```
Hmm one last thing. I find it a bit weird to do this over `fn sugg()` and return either `default_sugg`/`machine_applicable_sugg` plus the applicability.

It just feels, I dunno, a bit error prone to do `== MachineApplicable` 😅 and weird as usually applicability is lowered to `HasPlaceholders` if we fail to get a snippet of a span.

(Return type will be pretty complex tho. I think this is a sign we need some type alias for a multipart suggestion in the future...)

r=me after this

Centri3 replied:
PS the code is fine - I just think it should be rearranged a bit so that we don't match on `Applicability`. Having a fallback like this for when it specifically isn't machine applicable is oke

roife replied:
Merging the calculation of `sugg` and `applicability` into the same process would indeed make it more concise, then it would become:

```rust
fn sugg_appl(&self, parent_unwrap_call: &Option<Span>, primary_span: Span) -> (Applicability, Vec<(Span, String)>) {
        let Some(unwrap_span) = parent_unwrap_call else {
            return ...
        };

        match &self {
            FunctionKind::TryFromFunction(None) | FunctionKind::TryIntoFunction(None) => ...,
            _ => ...,
        }
    }
```

Do I need to make such a change?

Centri3 replied:
Yea smth like that.

> Do I need to make such a change?

I'd prefer so. tho I can merge it now and create a followup PR tomorrow once I have my desktop back

Centri3 commented:
@bors r+

ty for your work on this!

bors commented:
:pushpin: Commit 3c76b2ceff3526b2d46abb1bbd38180131578808 has been approved by `Centri3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=Centri3 3c76b2ceff3526b2d46abb1bbd38180131578808 -->
<!-- homu: {"type":"Approved","sha":"3c76b2ceff3526b2d46abb1bbd38180131578808","approver":"Centri3","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit 3c76b2ceff3526b2d46abb1bbd38180131578808 with merge 9c2373db9ed54dad6f87c212b168538b0db62e3b...
<!-- homu: {"type":"BuildStarted","head_sha":"3c76b2ceff3526b2d46abb1bbd38180131578808","merge_sha":"9c2373db9ed54dad6f87c212b168538b0db62e3b"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/7847206935/job/21415846309)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/actions/runs/7847206935/job/21415846309","builder_name":"checks-action_test"} -->

Centri3 commented:
https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Sporadic.20macos.20failures

@bors retry

bors commented:
:hourglass: Testing commit 3c76b2ceff3526b2d46abb1bbd38180131578808 with merge 28443e63fb633a5ed00a49e8df591a956d77122c...
<!-- homu: {"type":"BuildStarted","head_sha":"3c76b2ceff3526b2d46abb1bbd38180131578808","merge_sha":"28443e63fb633a5ed00a49e8df591a956d77122c"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/7847606273/job/21416916573), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/7847606279/job/21416920346), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/7847606277/job/21417152213)
Approved by: Centri3
Pushing 28443e63fb633a5ed00a49e8df591a956d77122c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Centri3","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7847606277/job/21417152213","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7847606273/job/21416916573","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7847606279/job/21416920346"},"merge_sha":"28443e63fb633a5ed00a49e8df591a956d77122c"} -->
